storybookjs/storybook
# Angular: Add multi-project setup for ng workspaces

Resolves https://github.com/storybookjs/storybook/issues/20213 
Resolves https://github.com/storybookjs/storybook/issues/14026 
Resolves https://github.com/storybookjs/storybook/issues/19289

## What I did

### Remove NX dependency

Previously, we tried to read the NX configuration during initialization to figure out which tsconfig we have to adjust for Storybook to work with Angular. We remove this kind of check. Instead, as soon as we recognize nx being part of the project, we exit the initialization and point to the excellent [@nrwl/storybook](https://nx.dev/packages/storybook) plugin instead.

### Removed support for Angular < 14

We do not support Angular below v14 anymore (https://github.com/storybookjs/storybook/pull/20559/commits/18b76181efb69d76ce7acf2c9c4aef275ff87e18)

### Remove the legacy way of using Storybook for Angular

In the past, it was possible to start Storybook independently of any [Angular builder](https://angular.io/guide/cli-builder). Running `yarn storybook` did the job. Now Storybook is bound to an Angular project, which is set up in the `angular.json` file. During initialization, new entries are added to a particular project's `architect` section to set up Storybook. You have to run `ng run <project>:storybook` and `ng run <project>:build-storybook` now instead. (https://github.com/storybookjs/storybook/pull/20559/commits/4b68d7558679099c1d5e37f4c883f1f34fbb6d65)

### Multi-project setup in Angular workspace

Before, one Storybook was set up for the entire Angular workspace. Since Storybook is now bound to a particular project, this is not the case anymore. If you have set up multiple applications/libraries within an Angular workspace, you will be asked for which particular project you would like to set up Storybook. (https://github.com/storybookjs/storybook/pull/20559/commits/542c5357c60ad63ce5b1c4d910a52a737a6ed410)

### Optional Compodoc generation

Previously, [compodoc](https://compodoc.app/) was set up automatically for you. You'll be asked if you'd like to set it up for a particular project.

### Fixes some Typescript-related issues related to Template Stories

The stricter set of Typescript rules was causing issues with the Template Stories, which are copied/linked over to a sandbox (https://github.com/storybookjs/storybook/pull/20559/commits/4f4b4e4501658d04db4768cb8eeb5365727b1c93)

### Removed legacy renderer

The `parameters.angularLegacyRendering` option is removed. You cannot use the old legacy renderer anymore.

## Flowchart

![F3A2D8DC-A76C-4B9D-A7DD-91869632F034_1_201_a](https://user-images.githubusercontent.com/5889929/211856070-88d36fb4-3083-4ec0-85b5-9883fe7c6fa7.jpeg)

## How to test

1. Create an angular sandbox (`yarn task --task sandbox --template angular-cli/default-ts --no-link`) (Script sometimes stops at `Publishing packages`. Rerun the command and start from `Run the internal npm server (run-registry)`.
2. `cd ./sandbox/angular-cli-default-ts`
3. `yarn storybook` -> Check whether Storybook runs smoothly. Check whether compodoc documentation is available in the addon section. Button inputs for example has some descriptions.
4. `npx ng generate application new-application` -> Creates a new application in the angular workspace
5. `<path-to-storybook>/code/lib/cli/bin/index.js init` -> Initialize Storybook for new-application. Answer compodoc prompt with `N`.
6. Go to angular.json and check whether `projects.angular-latest.architect.storybook` is configured with compodoc args and `projects.new-application.architect.storybook` is configured without compodoc args.
7. Go to `./projects/new-application/.storybook/main.js` and replace the content with:

```js
const mainRoot = require('../../../.storybook/main.js');
module.exports = {
  ...mainRoot,
  previewAnnotations: [],
  stories: ["../src/**/*.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"]
};
```

You must do this for testing because the sub-project does not have defined `template stories`. This adjustment, therefore, is only necessary due to how we set up sandboxes.

8. `npx ng run new-application:storybook`
9. `yarn build-storybook` -> Builds root project. Outputs Storybook dist to `storybook-static`
10. `npx ng run new-application:build-storybook` -> Builds new-application project. Outputs Storybook dist to `dist/storybook/new-application`
